### PR TITLE
Added rotating Icon

### PR DESCRIPTION
**Pull request by bastianleicht**
### Changes

- [x] New feature

Closes Issue: NaN 

### Description

This Commit adds some CSS, that rotate the Pin regenerate Icon when hovered.

### CODE REVIEW

**Comment by fbrettnich** [comment]
Hi, thanks for your PR
I think an animation when hovering makes little sense and is unnecessary. The animation would make more sense if the button is clicked until the new support PIN is there.
PR #3 will improve PIN regeneration using Ajax, this could well be added there..

**Comment by bastianleicht** [comment]
Hi, thanks for your Reply.
I think this is a good idea. I'm going to upate this PR, when #3 is merged.

**Comment by fbrettnich** [comment]
> Hi, thanks for your Reply. I think this is a good idea. I'm going to upate this PR, when #3 is merged.

FYI: PR #3 has been merged

**Comment by bastianleicht** [comment]
Alright, I'm going to update this PR in the next week.

**Comment by fbrettnich** [comment]
Hi, your new commits implement rotation on hover again.
The rotation should only start when clicking (before Ajax requests) and stop rotating after the new PIN is displayed (when Ajax is finished).

**Comment by bastianleicht** [comment]
Yes, I just updated my codespace. The next push is going to implement that feature.

**Comment by bastianleicht** [comment]
It now rotates the Icon while the Pin is generated.

File: modules/addons/supportpin/templates/client/index.tpl [hunk 1]
```diff
@@ -7,6 +7,13 @@
 *
 * Licensed under GPL-3.0 (https://github.com/fbrettnich/whmcs-supportpin-module/blob/main/LICENSE)
 -->
+<style>
+    .icon-rotate {
```
Comment (fbrettnich) [inline modules/addons/supportpin/templates/client/index.tpl:11]: I'm not sure about this, but normally Font Awesome should have the `fa-spin` class in it, right?

Comment (bastianleicht) [reply modules/addons/supportpin/templates/client/index.tpl:11]: Yes they have, but their implementation sometimes has a lot of problems. (Size, centering etc.) Here is a link to the Documentation.
https://fontawesome.com/v5.15/how-to-use/on-the-web/styling/animating-icons

**Comment by fbrettnich** [comment]
@ConanDoyl could you please submit a review?

**Comment by ConanDoyl** [comment]
> @ConanDoyl could you please submit a review?

Yes i'll check these changes and reply again if i'm done.

**Comment by ConanDoyl** [comment]
Seems good, but i would like to change one more thing to make it smoother. 
Let me show: 

Current Script:

```javascript
<script>
    let pinIcon = document.getElementById("pinIcon");
    function Request(url, callback){
        $.ajax({
           type: "POST",
           crossDomain: false,
           url: url,
           data: { "PIN": true },
           success: function(respond){
                callback(respond);
           },
           error: function(response){
               console.log(response.status);
           }
       });
    };
    
    function RenewPIN(){
        pinIcon.classList.add("icon-rotate");
        Request("index.php?m=supportpin&page=renew", function(response){
            $("#sPIN").html(response.PIN);
            pinIcon.classList.remove("icon-rotate");
        })
    }
    </script>
```

Let us update it to this one, i'll explain later why. 

```javascript

<script>
var isLoading = false;

function Request(url, callback, errorCallback = false) {
    $.ajax({
        type: "POST",
        crossDomain: false,
        url: url,
        data: { "PIN": true },
        success: function (respond) {
            callback(respond);
        },
        error: function (response) {
            if (!errorCallback) {
                console.log(response.status);
                return;
            }
            errorCallback(response);
        }
    });
};

function RenewPIN() {
    if (isLoading){
        return; // Break to avoid generating multiple Pins 
    }
    isLoading = true;
    $('#pinIcon').addClass('icon-rotate');
    Request("index.php?m=supportpin&page=renew", function (response) {
        $("#sPIN").html(response.PIN);
        $('#pinIcon').removeClass('icon-rotate');
        isLoading = false; // Remove the loading state again, if we add the button a ID we can update it also on Loadin with the attribute disabled and later on removing it again
    }, function (reponse) {
        // On Failure, we need to remove the rotating animation or if we also change the button, we need to remove this tho.
        isLoading = false;
        $('#pinIcon').removeClass('icon-rotate');
    });
}
    </script>
``` 

We using jQuery because WHMCS is using it also we do not need to afraid using it too. 
jQuery makes some things easier. Of course its not required. 
The second thing is, i've added is a loading State to check if the button has been pressed to avoid multiple post calls to the endpoint because it's absolutely not necessary. Then i've added a errorCallback to handle the failure event, if the request fails we change the rotating icon again back to static and make the user able to click it again.

**Review by ConanDoyl — DISMISSED** [review]
i can't find anything =)

File: modules/addons/supportpin/templates/client/index.tpl [hunk 2]
```diff
@@ -15,30 +22,47 @@
...
+            }
+        });
     };
     
-    function RenewPIN(){
-        Request("index.php?m=supportpin&page=renew", function(response){
+    function RenewPIN() {
+        if (isLoading){
+            return; // Break to avoid generating multiple Pins 
+        }
```
Comment (fbrettnich) [inline modules/addons/supportpin/templates/client/index.tpl]:
```suggestion
        if (isLoading) return;
```

File: modules/addons/supportpin/templates/client/index.tpl [hunk 3]
```diff
@@ -15,30 +22,47 @@
...
+        isLoading = true;
+        $('#pinIcon').addClass('icon-rotate');
+        Request("index.php?m=supportpin&page=renew", function (response) {
             $("#sPIN").html(response.PIN);
-        })
+            $('#pinIcon').removeClass('icon-rotate');
+            isLoading = false; // Remove the loading state again, if we add the button a ID we can update it also on Loadin with the attribute disabled and later on removing it again
```
Comment (fbrettnich) [inline modules/addons/supportpin/templates/client/index.tpl]:
```suggestion
            isLoading = false;
```

File: modules/addons/supportpin/templates/client/index.tpl [hunk 4]
```diff
@@ -15,30 +22,47 @@
...
+        }, function (reponse) {
+            // On Failure, we need to remove the rotating animation or if we also change the button, we need to remove this tho.
```
Comment (fbrettnich) [inline modules/addons/supportpin/templates/client/index.tpl]:
```suggestion
```

**Review by fbrettnich — DISMISSED** [review]

**Review by fbrettnich — APPROVED** [review]

**Pull request merged**